src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/TypeOperations.cs

```diff
@@ -2210,44 +2210,85 @@ private class CastInstructionT<T> : CastInstruction
             public override int Run(InterpretedFrame frame)
             {
                 var value = frame.Pop();
-                if (value != null)
-                {
-                    frame.Push((T)value);
-                }
-                else
-                {
-                    frame.Push(null);
-                }
+                frame.Push((T)value);
                 return +1;
             }
         }
 
-        private class CastInstructionNoT : CastInstruction
+        private abstract class CastInstructionNoT : CastInstruction
         {
             private readonly Type _t;
-            public CastInstructionNoT(Type t)
+            protected CastInstructionNoT(Type t)
             {
                 _t = t;
             }
 
+            public new static CastInstruction Create(Type t)
+            {
+                if (t.GetTypeInfo().IsValueType && !TypeUtils.IsNullableType(t))
+                {
+                    return new Value(t);
+                }
+                else
+                {
+                    return new Ref(t);
+                }
+            }
+
             public override int Run(InterpretedFrame frame)
             {
                 var value = frame.Pop();
                 if (value != null)
                 {
-                    if (!TypeUtils.HasReferenceConversion(value.GetType(), _t) &&
-                        !TypeUtils.HasIdentityPrimitiveOrNullableConversion(value.GetType(), _t))
+                    var valueType = value.GetType();
+
+                    if (!TypeUtils.HasReferenceConversion(valueType, _t) &&
+                        !TypeUtils.HasIdentityPrimitiveOrNullableConversion(valueType, _t))
+                    {
+                        throw new InvalidCastException();
+                    }
+
+                    if (!_t.IsAssignableFrom(valueType))
```

Comment:
@VSadov - Do you have any clue why we used HasReferenceConversion and HasIdentityPrimitiveOrNullableConversion without IsAssignableFrom? We need the latter for the negative cases, such as:

```
(System.Nullable`1[System.DayOfWeek]).Constant<System.Object>(1)
(System.Nullable`1[System.Int64]).Constant<System.Object>(1)
(System.String[]).Constant<System.Object[]>(System.Object[])
```

I'd like to make sure we're not overlooking anything that the two TypeUtils tests are correctly allowing but IsAssignableFrom is not, because we could now be throwing in more cases. On the other hand, just reasoning from a post-condition point of view, it wouldn't make much sense to push a value that can't be assigned to the desired target type (given that IsAssignableFrom is also used to check parameter/argument assignment compatibility etc.), so it seems IsAssignableFrom should return true for the push to make sense.